File: examples/basic/basic.c
/**
 * Copyright (C) 2015 Hewlett Packard Enterprise Development LP
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use self file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include "viface/viface.h"

// Interface configuration data
static char name[IFNAMSIZ] = "viface0";
static char ip[INET_ADDRSTRLEN] = "192.168.25.46";
static char mac[18] = "ec:f1:f8:d5:47:6b";
static char broadcast[INET_ADDRSTRLEN] = "192.168.25.255";
static char netmask[INET_ADDRSTRLEN] = "255.255.255.0";
static uint mtu = 1000;
static int id = 1;

/* Sets network configuration data
 * (name, id, ip, mac, netmask, broadcast, mtu).
 */
int set_interface_configuration(struct viface* self)
{
    // Configures interface
    if ((viface_set_ipv4(self, ip) == EXIT_FAILURE) ||
        (viface_set_mac(self, mac) == EXIT_FAILURE) ||
        (viface_set_ipv4_broadcast(self, broadcast) == EXIT_FAILURE) ||
        (viface_set_ipv4_netmask(self, netmask) == EXIT_FAILURE) ||
        (viface_set_mtu(self, mtu) == EXIT_FAILURE)) {
        return EXIT_FAILURE;
    }
    return EXIT_SUCCESS;
}

/* Checks the interface configuration data
 * (name, id, ip, mac, netmask, broadcast, mtu).
 */
int check_interface_configuration(struct viface* self)
{
    int id_value;
    bool is_viface_up;
    uint mtu_value;
    char* viface_name;
    char* ip_value;
    char* mac_value;
    char* broadcast_value;
    char* netmask_value;

    // Gets interface configuration data
    if ((viface_get_name(self, &viface_name) == EXIT_FAILURE) ||
        (viface_get_id(self, &id_value) == EXIT_FAILURE) ||
        (viface_is_up(self, &is_viface_up) == EXIT_FAILURE) ||
        (viface_get_ipv4(self, &ip_value) == EXIT_FAILURE) ||
        (viface_get_mac(self, &mac_value) == EXIT_FAILURE) ||
        (viface_get_ipv4_broadcast(self, &broadcast_value) == EXIT_FAILURE) ||
        (viface_get_ipv4_netmask(self, &netmask_value) == EXIT_FAILURE) ||
        (viface_get_mtu(self, &mtu_value) == EXIT_FAILURE)) {
        return EXIT_FAILURE;
    }

    printf("********** Viface Configuration Data **********\n\n");

    // Checks interface name
    printf("--- Viface name is: %s.\n", viface_name);
    printf("    Expected:       %s.\n\n", name);

    // Checks interface ID
    printf("--- Viface ID is: %d.\n", id_value);
    printf("    Expected:     %d.\n\n", id);

    // Checks if interface is Up
    printf("--- Interface isUp: %s\n", is_viface_up ? "true" : "false");
    printf("    Expected:       true.\n\n");

    // Checks interface IP
    printf("--- IP value is: %s\n", ip_value);
    printf("    Expected:    %s.\n\n", ip);

    // Checks mac value
    printf("--- MAC value is: %s\n", mac_value);
    printf("    Expected:     %s.\n\n", mac);

    // Checks broadcast value
    printf("--- Broadcast value is: %s\n", broadcast_value);
    printf("    Expected:           %s.\n\n", broadcast);

    // Checks netmask value
    printf("--- Netmask value is: %s\n", netmask_value);
    printf("    Expected:         %s.\n\n", netmask);

    // Checks mtu value
    printf("--- MTU value is: %d\n", mtu_value);
    printf("    Expected:     %u.\n\n", mtu);

    return EXIT_SUCCESS;
}

// Checks if the network interface is down
int check_interface_is_down(struct viface* self)
{
    // Checks if interface is down
    bool is_down;

    if (viface_is_up(self, &is_down) == EXIT_FAILURE) {
        return EXIT_FAILURE;
    }

    printf("--- Interface isDown: %s\n", is_down ? "false" : "true");
    printf("    Expected:         true\n\n");
    return EXIT_SUCCESS;
}

/**
 * This example shows the basics to setup a new virtual network
 * interface called viface0
 * (or use the name passed as first argument).
 * It will show how to:
 *  1) Set interface configuration data
 *     (name, id, ip, ipv6, mac, netmask, broadcast, mtu).
 *  2) Bring-up the network interface.
 *  3) Check the interface configuration data
 *     (name, id, ip, ipv6, mac, netmask, broadcast, mtu).
 */
main(int argc, const char* argv[])
{
    struct viface* self;

    printf("\n--- Starting basic example...\n\n");

    if (argc > 1) {
        strcpy(name, argv[1]);
    }

    /* These IF statements do the following:
     * 1) Creates global parent APR pool
     * 2) Creates interface
     * 3) Sets interface configuration data
     * 4) Brings-up interface
     * 5) Checks interface configuration data
     * 6) Brings-down interface
     * 7) Checks if interface is down
     * 8) Destroys viface struct
     * 9) Destroys global APR pool
     */
    if ((viface_create_global_pool() == EXIT_FAILURE) ||
        (viface_create_viface(name, true, id, &self) == EXIT_FAILURE) ||
        (set_interface_configuration(self) == EXIT_FAILURE) ||
        (viface_up(self) == EXIT_FAILURE) ||
        (check_interface_configuration(self) == EXIT_FAILURE) ||
        (viface_down(self) == EXIT_FAILURE) ||
        (check_interface_is_down(self) == EXIT_FAILURE) ||
        (viface_destroy_viface(&self) == EXIT_FAILURE) ||
        (viface_destroy_global_pool() == EXIT_FAILURE)) {
        return EXIT_FAILURE;
    }

    return 0;
}

File: bindings/go/src/examples/basic/basic.go
package main

import (
    "fmt"
    "vifacego"
)

// Viface configuration data
var name = "viface0"
var ipv4 = "192.168.25.46"
var ipv6s = []string{"fe80::6829:26ff:fe53:4670",
                     "fe20::5478:93f5:fe93:4680"}
var mac = "ec:f1:f8:d5:47:6b"
var broadcast = "192.168.25.255"
var netmask = "255.255.255.0"
var id = 1
var mtu uint = 1500 //Default value if not set

// setInterfaceConfiguration Sets network configuration data
// (name, id, ip, mac, netmask, broadcast, mtu).
func setInterfaceConfiguration(self vifacego.Viface) int {
    // Configures interface
    if ((vifacego.VifaceSetIpv4(self, ipv4) == vifacego.ExitFailure) ||
        (vifacego.VifaceSetMac(self, mac) == vifacego.ExitFailure) ||
        (vifacego.VifaceSetIpv4Broadcast(self,broadcast) ==
            vifacego.ExitFailure) ||
        (vifacego.VifaceSetIpv4Netmask(self, netmask) ==
            vifacego.ExitFailure) ||
        (vifacego.VifaceSetIpv6(self, len(ipv6s), ipv6s) ==
            vifacego.ExitFailure)) {
        return vifacego.ExitFailure
    }

    return vifacego.ExitSuccess;
}

// checkInterfaceConfiguration Checks the interface configuration data
// (name, id, ip, mac, netmask, broadcast, mtu).
func checkInterfaceConfiguration(self vifacego.Viface) int {
    var idValue uint
    var vifaceUp bool
    var mtuValue uint
    var vifaceName string
    var ipValue string
    var macValue string
    var broadcastValue string
    var netmaskValue string
    var ipv6sValues []string

    // Gets interface configuration data
    if ((vifacego.VifaceGetName(self, &vifaceName) == vifacego.ExitFailure) ||
        (vifacego.VifaceGetID(self, &idValue) == vifacego.ExitFailure) ||
        (vifacego.VifaceIsUp(self, &vifaceUp) == vifacego.ExitFailure) ||
        (vifacego.VifaceGetIpv4(self, &ipValue) == vifacego.ExitFailure) ||
        (vifacego.VifaceGetMac(self, &macValue) == vifacego.ExitFailure) ||
        (vifacego.VifaceGetIpv4Broadcast(self, &broadcastValue) ==
            vifacego.ExitFailure) ||
        (vifacego.VifaceGetIpv4Netmask(self, &netmaskValue) ==
            vifacego.ExitFailure) ||
        (vifacego.VifaceGetMtu(self, &mtuValue) == vifacego.ExitFailure) ||
        (vifacego.VifaceGetIpv6(self, &ipv6sValues) == vifacego.ExitFailure)) {
        return vifacego.ExitFailure
    }

    fmt.Printf("********** Viface Configuration Data **********\n\n");

    // Checks interface name
    fmt.Printf("--- Viface name is: %v.\n", vifaceName);
    fmt.Printf("    Expected:       %v.\n\n", name);

    // Checks interface ID
    fmt.Printf("--- Viface ID is: %v.\n", idValue);
    fmt.Printf("    Expected:     %v.\n\n", id);

    // Checks if interface is Up
    fmt.Printf("--- Interface isUp: %v\n", vifaceUp);
    fmt.Printf("    Expected:       true.\n\n");

    // Checks interface IP
    fmt.Printf("--- IP value is: %v\n", ipValue);
    fmt.Printf("    Expected:    %v.\n\n", ipv4);

    // Checks mac value
    fmt.Printf("--- MAC value is: %v\n", macValue);
    fmt.Printf("    Expected:     %v.\n\n", mac);

    // Checks broadcast value
    fmt.Printf("--- Broadcast value is: %v\n", broadcastValue);
    fmt.Printf("    Expected:           %v.\n\n", broadcast);

    // Checks netmask value
    fmt.Printf("--- Netmask value is: %v\n", netmaskValue);
    fmt.Printf("    Expected:         %v.\n\n", netmask);

    // Checks mtu value
    fmt.Printf("--- MTU value is: %v\n", mtuValue);
    fmt.Printf("    Expected:     %v.\n\n", mtu);

    // Checks Ipv6s values
    fmt.Printf("--- Ipv6 values are:\n");

    for i:= 0; i < len(ipv6sValues); i++ {
        fmt.Printf("        %v\n", ipv6sValues[i]);
    }
    fmt.Printf("\n");

    return vifacego.ExitSuccess
}

// checkInterfaceIsDown Checks if the network interface is down
func checkInterfaceIsDown(self vifacego.Viface) int {
    // Checks if interface is down
    var isDown bool

    if (vifacego.VifaceIsUp(self, &isDown) == vifacego.ExitFailure) {
        return vifacego.ExitFailure;
    }

    fmt.Printf("--- Interface is still Up: %v\n", isDown);
    fmt.Printf("    Expected:              false\n\n");
    return vifacego.ExitSuccess;
}


// main This example shows the basics to setup a new virtual network
// interface called viface0
// (or use the name passed as first argument).
// It will show how to:
//  1) Set interface configuration data
//     (name, id, ip, ipv6, mac, netmask, broadcast, mtu).
//  2) Bring-up the network interface.
//  3) Check the interface configuration data
//     (name, id, ip, ipv6, mac, netmask, broadcast, mtu).
func main() {

    fmt.Printf("\n--- Starting basic GO example...\n\n")

    // Creates viface struct
    var self = vifacego.CreateVifaceStruct()

    /* These IF statements do the following:
     * 1) Creates global parent APR pool
     * 2) Creates interface
     * 3) Sets interface configuration data
     * 4) Brings-up interface
     * 5) Checks interface configuration data
     * 6) Brings-down interface
     * 7) Checks if interface is down
     * 8) Destroys viface struct
     * 9) Destroys global APR pool
     */
    if ((vifacego.VifaceCreateGlobalPool() == vifacego.ExitFailure) ||
        (vifacego.VifaceCreateViface(name, true, id, &self) ==
            vifacego.ExitFailure) ||
        (setInterfaceConfiguration(self) == vifacego.ExitFailure) ||
        (vifacego.VifaceUp(self) == vifacego.ExitFailure) ||
        (checkInterfaceConfiguration(self) == vifacego.ExitFailure) ||
        (vifacego.VifaceDown(self) == vifacego.ExitFailure) ||
        (checkInterfaceIsDown(self) == vifacego.ExitFailure) ||
        (vifacego.VifaceDestroyViface(self) == vifacego.ExitFailure) ||
        (vifacego.VifaceDestroyGlobalPool() == vifacego.ExitFailure)) {
        fmt.Printf("--- An error occurred executing basic GO example\n")
    }


}
